server/src/main/java/org/opensearch/indices/replication/OngoingSegmentReplications.java

```diff
@@ -0,0 +1,195 @@
+        if (nodesToHandlers.containsKey(node)) {
+            final SegmentReplicationSourceHandler handler = nodesToHandlers.get(node);
+            // update the given listener to release the CopyState before it resolves.
+            final ActionListener<GetSegmentFilesResponse> wrappedListener = ActionListener.runBefore(listener, () -> {
+                final SegmentReplicationSourceHandler sourceHandler = nodesToHandlers.remove(node);
+                removeCopyState(sourceHandler.getCopyState());
+            });
+            handler.sendFiles(request, wrappedListener);
+        } else {
+            listener.onResponse(new GetSegmentFilesResponse(Collections.emptyList()));
+        }
+    }
+
+    /**
+     * Remove a CopyState. Intended to be called after a replication event completes.
+     * This method will remove a copyState from the copyStateMap only if its refCount hits 0.
+     * @param copyState {@link CopyState}
+     */
+    private synchronized void removeCopyState(CopyState copyState) {
+        copyState.decRef();
+        if (copyState.refCount() <= 0) {
+            copyStateMap.remove(copyState.getRequestedReplicationCheckpoint());
+        }
+    }
+
+    /**
+     * Prepare for a Replication event. This method constructs a {@link CopyState} holding files to be sent off of the current
+     * nodes's store.  This state is intended to be sent back to Replicas before copy is initiated so the replica can perform a diff against its
+     * local store.  It will then build a handler to orchestrate the segment copy that will be stored locally and started on a subsequent request from replicas
+     * with the list of required files.
+     * @param request {@link CheckpointInfoRequest}
+     * @param fileChunkWriter {@link FileChunkWriter} writer to handle sending files over the transport layer.
+     * @return {@link CopyState} the built CopyState for this replication event.
+     * @throws IOException - When there is an IO error building CopyState.
+     */
+    CopyState prepareForReplication(CheckpointInfoRequest request, FileChunkWriter fileChunkWriter) throws IOException {
+        final CopyState copyState = getCachedCopyState(request.getCheckpoint());
+        final SegmentReplicationSourceHandler handler = createTargetHandler(request.getTargetNode(), copyState, fileChunkWriter);
+        nodesToHandlers.putIfAbsent(request.getTargetNode(), handler);
+        return copyState;
+    }
+
+    int size() {
+        return nodesToHandlers.size();
+    }
+
+    int cachedCopyStateSize() {
+        return copyStateMap.size();
+    }
+
+    /**
+     * Cancel all Replication events for the given shard, intended to be called when the current primary is shutting down.
+     * @param shard {@link IndexShard}
+     * @param reason  {@link String} - Reason for the cancel
+     */
+    public void cancel(IndexShard shard, String reason) {
+        for (SegmentReplicationSourceHandler entry : nodesToHandlers.values()) {
```

Comment:
AFAIK, making Map synchronized does not guarantee safe iteration, as per documentation [1]:

> It is imperative that the user manually synchronize on the returned map when iterating over any of its collection views:

Please make `cancel` synchronized or use `ConcurrentMap` for `nodesToHandlers`

[1] https://docs.oracle.com/javase/7/docs/api/java/util/Collections.html#synchronizedMap%28java.util.Map%29